### DISCUSSION TITLE
http3 support for production

### DISCUSSION THREAD

**Discussion by asman1337** [salvo-rs/salvo#279] [Q&A]
I enabled http3 support now getting error in production.

server running normally -
`server[29015]: 2023-05-08T13:12:19.259289Z  INFO salvo_core::server: listening HTTP/3.0 on https://127.0.0.1:5800
 server[29015]: 2023-05-08T13:12:19.259314Z  INFO salvo_core::server: listening HTTP/2.0 on https://127.0.0.1:5800
 server[29015]: 2023-05-08T13:12:19.260017Z  INFO salvo_core::conn::rustls::listener: tls config loaded.`

but When trying to open website, it is throwing error -
'ERROR salvo_core::server: http serve connection failed error=Custom { kind: Other,error: "received corrupt message" }'

here is my main function - 

`#[tokio::main]
async fn main() {
    tracing_subscriber::fmt().init();

    /* Setting up certificate for quic */
    let cert = include_bytes!("../../certs/cert.pem").to_vec();
    let key = include_bytes!("../../certs/key.pem").to_vec();

    /* CORS Config. */
    let _cors_handler = Cors::new()
        .allow_origin(BASE_URL)
        .allow_methods(vec![Method::GET, Method::POST, Method::DELETE])
        .into_handler();

    /* Configure routes to be served */
    let router = Router::new().get(index);
    let config = RustlsConfig::new(
        Keycert::new()
            .with_cert(cert.as_slice())
            .with_key(key.as_slice()),
    );

    /* Configuring and Starting the server. */
    let listener = TcpListener::new(("127.0.0.1", SERVER_PORT)).rustls(config.clone());
    let acceptor = QuinnListener::new(config, ("127.0.0.1", SERVER_PORT))
        .join(listener)
        .bind()
        .await;

    Server::new(acceptor).serve(router).await;
}`

I have a folder named `certs` in my root directory and inside cert.pem and key.pem. But server returning bad request.
or is there any way I can directly user certbot/lets encrypt certificate?

**Comment by asman1337** [comment]
@chrislearn Bro what is the correct way to pass AcmeListener to QuinnListener?
I didn't find any such documention either, or How can I  configure QuinnListener to use LetsEncrypt certificate.
I am using Caddy Server, In that case It is already serving over ssl by default.

**Comment by chrislearn** [comment]
There is a `hello-h3` example in examples folder, is it worked?

**Comment by asman1337** [comment]
@chrislearn Yes bro I am following `hello-h3` and It's working fine locally but getting above mentioned error in production server.
Maybe because of self-signed openssl keys.
But I didn't find a way either to use acme with `QuinnListener` , if you can provide any sample on how to use acme with QuinnListener.


Server OS : Ubuntu 22.04.2 LTS
WebServer : Caddy - HTTP/1-2-3 web server with automatic HTTPS
`Cloudflare SSL/TLS encryption mode is Full`

Server is running but when sending get to `/` root or opening url in browser, getting following error -
`'ERROR salvo_core::server: http serve connection failed error=Custom { kind: Other,error: "received corrupt message" }'`

**Comment by zhuravkov** [comment]
I also could not figure out how to run h3 in production on port 443, or issue http2 (unless let listener = TcpListener::new(("0.0.0.0", 443)).rustls(config.clone()); and let acceptor = QuinnListener::new(config, ("0.0.0.0", 80))) or nothing at all! It would be nice to see an example of running on the master port with letsencrypt

```
#[tokio::main]
async fn main() {
    tracing_subscriber::fmt().init();
    let cert = include_bytes!("../acme/letsencrypt/cert.pem").to_vec();
    let key = include_bytes!("../acme/letsencrypt/key.pem").to_vec();

    let router = Router::new().get(hello);
    let config = RustlsConfig::new(Keycert::new().cert(cert.as_slice()).key(key.as_slice()));
    let listener = TcpListener::new(("0.0.0.0", 443)).rustls(config.clone());

    let acceptor = QuinnListener::new(config, ("0.0.0.0", 443))
        .join(listener)
        .bind()
        .await;

    Server::new(acceptor).serve(router).await;
}
```
2023-08-27T17:16:45.666619Z  INFO salvo_core::server: listening [HTTP/3.0] on https://0.0.0.0:443
2023-08-27T17:16:45.666787Z  INFO salvo_core::server: listening [HTTP/1.1, HTTP/2.0] on https://0.0.0.0:443
2023-08-27T17:16:45.666929Z  INFO salvo_core::conn::rustls::listener: tls config loaded.

**Comment by chrislearn** [comment]
Acme work with Quinn examples:
https://github.com/salvo-rs/salvo/blob/main/examples/acme-http01-quinn/src/main.rs
https://github.com/salvo-rs/salvo/blob/main/examples/webtransport-acme-http01/src/main.rs

**Comment by zhuravkov** [comment]
```
23 |       let listener = TcpListener::new("0.0.0.0:443")
   |  ____________________-
24 | |         .acme()
25 | |         .cache_path("cert")
26 | |         .add_domain("domain.com")
27 | |         .http01_challege(&mut router).quinn("0.0.0.0:443");
   | |                                      -^^^^^ method not found in `AcmeListener<TcpListener<&str>>`
```

error in example https://github.com/salvo-rs/salvo/blob/main/examples/acme-http01-quinn/src/main.rs

**Comment by chrislearn** [comment]
make sure you are using latest salvo 0.52.0
> From: ***@***.***>
> Date:  Tue, Aug 29, 2023, 03:07
> Subject:  Re: [salvo-rs/salvo] http3 support for production (Discussion #279)
> To: ***@***.***>
> Cc: "Chrislearn ***@***.***>, ***@***.***>
> 23 |       let listener = TcpListener::new("0.0.0.0:443")
   |  ____________________-
24 | |         .acme()
25 | |         .cache_path("cert")
26 | |         .add_domain("domain.com<http://domain.com>")
27 | |         .http01_challege(&mut router).quinn("0.0.0.0:443");
   | |                                      -^^^^^ method not found in `AcmeListener<TcpListener<&str>>`
error in example https://github.com/salvo-rs/salvo/blob/main/examples/acme-http01-quinn/src/main.rs

**Comment by zhuravkov** [comment]
made update to salvo 0.52.0
```
    cfg_feature! {
        #![feature = "quinn"]
```
returns AcmeQuinnListener::new(self, local_addr)   it is like hello h3 example!
Now
 2023-08-28T20:16:23.138104Z  INFO salvo_core::server: listening [HTTP/1.1, HTTP/2.0] on https://0.0.0.0:443
2023-08-28T20:16:23.138346Z  INFO salvo_core::server: listening [HTTP/3.0] on https://0.0.0.0:443
2023-08-28T20:16:23.138405Z  INFO salvo_core::server: listening [HTTP/1.1] on http://0.0.0.0:80
 in f12/network/protocol is h2
i think h2 works before h3,

**Comment by chrislearn** [comment]
http3 is upgraded from http2. You can use firefox and refresh page. protocol will change to http3.
> From: ***@***.***>
> Date:  Tue, Aug 29, 2023, 04:31
> Subject:  Re: [salvo-rs/salvo] http3 support for production (Discussion #279)
> To: ***@***.***>
> Cc: "Chrislearn ***@***.***>, ***@***.***>
> made update to salvo 0.52.0    cfg_feature! {
        #![feature = "quinn"]
returns AcmeQuinnListener::new(self, local_addr)   it is like hello h3 example!
> 
Now
> 
2023-08-28T20:16:23.138104Z  INFO salvo_core::server: listening [HTTP/1.1, HTTP/2.0] on https://0.0.0.0:443
> 
2023-08-28T20:16:23.138346Z  INFO salvo_core::server: listening [HTTP/3.0] on https://0.0.0.0:443
> 
2023-08-28T20:16:23.138405Z  INFO salvo_core::server: listening [HTTP/1.1] on http://0.0.0.0:80
> 
in f12/network/protocol is h2
> 
i think h2 works before h3,
